instantsearch/src/main/java/com/algolia/instantsearch/ui/InstantSearch.java

```diff
@@ -289,53 +285,89 @@ private void processActivity(@NonNull final Activity activity) {
             searchView.setSearchableInfo(manager.getSearchableInfo(activity.getComponentName()));
         }
 
-        // Register any AlgoliaWidget
-        final List<AlgoliaWidget> foundListeners = LayoutViews.findByClass(rootView, AlgoliaWidget.class);
-        if (foundListeners.size() == 0) {
-            throw new IllegalStateException(Errors.LAYOUT_MISSING_HITS);
-        }
         final List<String> refinementAttributes = new ArrayList<>();
-        for (AlgoliaWidget widget : foundListeners) {
-            processWidget(rootView, widget, refinementAttributes);
+
+        // Register any AlgoliaResultListener
+        final List<AlgoliaResultListener> resultListeners = LayoutViews.findByClass(rootView, AlgoliaResultListener.class);
+        if (resultListeners.size() == 0) {
+            throw new IllegalStateException(Errors.LAYOUT_MISSING_RESULT_LISTENER);
+        }
+        for (AlgoliaResultListener listener : resultListeners) {
+            processResultListener(rootView, listener, refinementAttributes);
+        }
+
+        // Register any AlgoliaErrorListener
+        final List<AlgoliaErrorListener> errorListeners = LayoutViews.findByClass(rootView, AlgoliaErrorListener.class);
+        for (AlgoliaErrorListener listener : errorListeners) {
+            processErrorListener(rootView, listener, refinementAttributes);
+        }
+
+        // Register any AlgoliaSearcherListener
+        final List<AlgoliaSearcherListener> searcherListeners = LayoutViews.findByClass(rootView, AlgoliaSearcherListener.class);
+        for (AlgoliaSearcherListener listener : searcherListeners) {
+            processSearcherListener(rootView, listener, refinementAttributes);
         }
+
```

Comment:
refactor nit: you can extract L290-L309 into a method processAllListeners(rootView, listener, refinementAttributes)